### PR TITLE
[SPARK-12837][CORE] reduce network IO for accumulators

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Sending un-updated accumulators back to driver makes no sense, as merging a zero value accumulator is a no-op. We should only send back updated accumulators, to save network IO.
## How was this patch tested?

new test in `TaskContextSuite`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @davies @rxin

**Comment by yhuai** [comment]
test this please

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -1097,8 +1097,8 @@ class DAGScheduler(
             throw new SparkException(s"attempted to access non-existent accumulator $id")
         }
         acc.merge(updates.asInstanceOf[AccumulatorV2[Any, Any]])
-        // To avoid UI cruft, ignore cases where value wasn't updated
-        if (acc.name.isDefined && !updates.isZero) {
+        // Only display named accumulators on UI.
+        if (acc.name.isDefined) {
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Why this change?

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala [hunk 2]
```diff
@@ -291,25 +291,32 @@ private[spark] object TaskMetrics extends Logging {
 
 private[spark] class BlockStatusesAccumulator
   extends AccumulatorV2[(BlockId, BlockStatus), Seq[(BlockId, BlockStatus)]] {
-  private[this] var _seq = ArrayBuffer.empty[(BlockId, BlockStatus)]
+  private[this] var _seq: ArrayBuffer[(BlockId, BlockStatus)] = _
 
-  override def isZero(): Boolean = _seq.isEmpty
+  private def seq = {
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: Should this be thread-safe?

**Comment by davies** [comment]
@cloud-fan How much we can gain from 2)?

**Comment by davies** [comment]
```
scala> ser.newInstance().serialize(ArrayBuffer.empty[Any])
res4: java.nio.ByteBuffer = java.nio.HeapByteBuffer[pos=0 lim=173 cap=214]

scala> ser.newInstance().serialize(null)
res5: java.nio.ByteBuffer = java.nio.HeapByteBuffer[pos=0 lim=5 cap=32]

scala> ser.newInstance().serialize(new java.util.ArrayList[Long])
res6: java.nio.ByteBuffer = java.nio.HeapByteBuffer[pos=0 lim=58 cap=64]

scala> ser.newInstance().serialize(1L)
res7: java.nio.ByteBuffer = java.nio.HeapByteBuffer[pos=0 lim=82 cap=128]

scala> ser.newInstance().serialize(Array(1L))
res8: java.nio.ByteBuffer = java.nio.HeapByteBuffer[pos=0 lim=35 cap=64]
```

**Comment by rxin** [comment]
@cloud-fan we don't need to do this here but I think we can also substantially cut down the size of a task result if we consolidate all the accumulators into a single one in TaskMetrics.

**Comment by andrewor14** [comment]
I agree. I think (1) is straightforward and we should do it. (2) I'm not so sure since it only affects one of the accumulators.

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala (hunk 2 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: can you add return type

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: it's because the executor no longer sends back updates where the value is zero, so the second condition is always assumed to be true. Maybe we should add an assert or something instead.

**Comment by davies** [comment]
Since we have only one BlockStatusesAccumulator object in TaskMetrics, it may not worth to do 2).

**Comment by davies** [comment]
I will be great we could have a test to make sure that we always have this behavior.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
@davies , I think the added `assert` can guarantee this?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 1 above)
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: I reverted this change because we can't do assert here. It's surrounded by a `try catch` which captures all NonFatal exception.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/Task.scala [hunk 3]
```diff
@@ -155,7 +155,13 @@ private[spark] abstract class Task[T](
    */
   def collectAccumulatorUpdates(taskFailed: Boolean = false): Seq[AccumulatorV2[_, _]] = {
     if (context != null) {
-      context.taskMetrics.accumulators().filter { a => !taskFailed || a.countFailedValues }
+      context.taskMetrics.internalAccums.filter { a =>
+        // RESULT_SIZE accumulator is always zero at executor, we need to send it back as its
+        // value will be updated at driver side.
+        !a.isZero || a.name == Some(InternalAccumulator.RESULT_SIZE)
+      // zero value external accumulators may still be useful, e.g. SQLMetrics, we should not filter
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: We change the countOnFailure from false to true recently, is that an design change?

Why SQLMetrics are external accumulators?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: internal here means task metrics. "internal" to core.

Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: There are 2 concepts:

1. internal accumulators: like GCtime, resultSize, which are internal to DAGScheduler.
2. `countFailedValues` accumulator: `countFailedValues` is an internal flag that can only be set by us. All internal accumulators are `countFailedValues` accumulators, and SQLMetrics are also `countFailedValues` accumulators.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: we should probably change the semantics of internal to mean internal to Spark (i.e. include SQL metrics), but that's a separate issue.

**Comment by andrewor14** [comment]
This patch itself LGTM. I'm merging it into master 2.0

File: core/src/main/scala/org/apache/spark/scheduler/Task.scala (hunk 3 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: SQL metrics should not count on failures, will we fix that in this PR or a separate one? Then this part should also be updated.

Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: Another question is that: Why zero value of SQL metrics are useful? Are they only useful for tests or UI?

Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: Because SQL Metrics has some statistics, e.g. max, min, avg, we need all values even it's zero.

Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: and it's for UI.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 4]
```diff
@@ -66,7 +66,7 @@ private[sql] object SQLMetrics {
 
   def createMetric(sc: SparkContext, name: String): SQLMetric = {
     val acc = new SQLMetric(SUM_METRIC)
-    acc.register(sc, name = Some(name), countFailedValues = true)
+    acc.register(sc, name = Some(name), countFailedValues = false)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: It was a mistake, SQLMetric should not count failed values.

File: core/src/main/scala/org/apache/spark/scheduler/Task.scala (hunk 3 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: We could see zero value as `null`, it's reasonable to not include the `null` in sum/avg. Is there any downside to exclude these zeros ?

Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: It's task level statistics. Let's say an operator launched 100 tasks to execute it, and 99 tasks don't spill, only one task spills 100 mb, then the avg of `spilling size` will be 100 mb, if we don't include zero values. This is obviously wrong.

Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: If there is no spilling, we could say the size of spilling is not define (null for unknown). We also have total value, we could know that how many task had spilled. Right now, we can't know how many tasks had spilled, actually it's worse.

I don't think it's wrong.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: If this is the only issue, can we merge this pull request and have a new pr to fix this semantics - which clearly there are some disagreement with.

Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: Sounds good to me.

File: core/src/main/scala/org/apache/spark/scheduler/Task.scala [hunk 5]
```diff
@@ -155,7 +155,13 @@ private[spark] abstract class Task[T](
    */
   def collectAccumulatorUpdates(taskFailed: Boolean = false): Seq[AccumulatorV2[_, _]] = {
     if (context != null) {
-      context.taskMetrics.accumulators().filter { a => !taskFailed || a.countFailedValues }
+      context.taskMetrics.internalAccums.filter { a =>
+        // RESULT_SIZE accumulator is always zero at executor, we need to send it back as its
+        // value will be updated at driver side.
+        !a.isZero || a.name == Some(InternalAccumulator.RESULT_SIZE)
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: Can we add a comment to say internal accumulators will always count on failures?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: ok I'll add it when I merge

**Comment by andrewor14** [comment]
LGTM merging into master 2.0. Let's address any follow-ups in a future patch.
